pkg/cmd/search/shared/searcher.go

```diff
@@ -0,0 +1,169 @@
+package shared
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/cli/cli/v2/pkg/search"
+)
+
+type searcher struct {
+	client *http.Client
+	host   string
+}
+
+func NewSearcher(host string, client *http.Client) *searcher {
+	return &searcher{
+		client: client,
+		host:   host,
+	}
+}
+
+func (s *searcher) Search(query search.Query) (search.Result, error) {
```

Comment:
Right now this does not support any sort of parallelization or progress indication through returning results as they are retrieved. I thought I would leave these as improvements for the future.